src/include/daos/checksum.h

```diff
@@ -15,48 +15,259 @@
  *
  * GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
  * The Government's rights to use, modify, reproduce, release, perform, display,
- * or disclose this software are subject to the terms of the Apache License
+ * or disclose this software are subject to the terms of the Apache License as
  * provided in Contract No. B609815.
  * Any reproduction of computer software, computer software documentation, or
  * portions thereof marked with this legend must also reproduce the markings.
  */
 
-#ifndef __DAOS_CHECKSUM_H
-#define __DAOS_CHECKSUM_H
+#ifndef __DAOS_CHECKSUMMER_H
+#define __DAOS_CHECKSUMMER_H
 
-#include <daos/common.h>
-#if defined(__x86_64__)
-#include <isa-l.h>
-#else
-#include <mchecksum.h>
-#endif
+#include <daos_types.h>
+#include <daos_obj.h>
+#include <daos_prop.h>
 
-#define DAOS_CSUM_SIZE 64
+/**
+ * -----------------------------------------------------------
+ * Container Property Knowledge
+ * -----------------------------------------------------------
+ */
+uint32_t
+daos_cont_prop2csum(daos_prop_t *props);
+
+uint64_t
+daos_cont_prop2chunksize(daos_prop_t *props);
+
+bool
+daos_cont_csum_prop_is_valid(uint16_t val);
+
+bool
+daos_cont_csum_prop_is_enabled(uint16_t val);
+
+/**
+ * -----------------------------------------------------------
+ * DAOS Checksummer
+ * -----------------------------------------------------------
+ */
+/** Type of checksums DAOS supports. Primarily used for looking up the
+ * appropriate algorithm functions to be used for the csummer
+ */
+enum DAOS_CSUM_TYPE {
+	CSUM_TYPE_UNKNOWN = 0,
+
+	CSUM_TYPE_ISAL_CRC16_T10DIF = 1,
```

Comment:
I would still keep the type generic and pick based on the best possible checksum available on the system. It would better to have that control within the checksum library than outside. I am worried this will cause an enum proliferation internally.

Reply:
I think picking which checksum is still kept within the checksum library. My intention with this enum is to have one per algorithm supported so the algorithm function tables can stay hidden in checksum.c. Users will create a container with a container checksum type property which is more generic (i.e. DAOS_PROP_CO_CSUM_CRC16). Then daos_contprop2csumtype (internal to the checksum library) is used internally to get which checksum algorithm to use, represented by the DAOS_CSUM_TYPE. Right now daos_contprop2csumtype simply uses a switch/case, but I guess more logic can be added as needed. 
Maybe we should rename DAOS_CSUM_TYPE to something like CSUM_ALGO?